### ISSUE TITLE
Kernel: Deadlock on SpinLock in WaitBlocker::add_blocker from waitid

### ISSUE THREAD

**Issue by ADKaster** [SerenityOS/serenity#9540]
Setup: I was running the clang i686 build with UBSAN enabled, and KVM disabled. 

For WSL I couldn't convince it not to set the env var so I just commented out the SERENITY_VIRT_TECH_ARG lines in run.sh

Booted into text mode with ` ./Meta/serenity.sh gdb i686 boot_mode=no-fbdev`

I normally run with SERENITY_RAM_SIZE=1G so I made sure to unset that this time.

```
# su anon
~ cd /home/anon
~ for i in {1..50} { run-tests /usr/Tests/Shell }
```
And then wait ... I got the deadlock at around process 6038, in  the sigpipe.sh test.

![blocked](https://user-images.githubusercontent.com/8388494/130311548-62b9c97f-eb59-4908-b718-f93066c8c798.PNG)

I saw from the backtrace that it was inside the WaitBlocker constructor in do_waitid, calling Thread::set_block_condition, then BlockCondition::add_blocker, which takes the BlockConditon's spinlock. 

I tried to print out the list of runnable threads, but seems that was a bad plan as it got a page fault doing so and panic'd

```
[#0 yes(6005:6005)]: Terminating yes(6005) due to signal 15
[#0 Shell(6038:6038)]: Scheduler thread list for processor 0:
[Shell(6038:6038)]:   Runnable       Pr: 1 colonel(0:0)                   @ 0008:c065426c (nsched: 378800)
[Shell(6038:6038)]:                  Pr:30 IO WorkQueue(2:2)              @ 0008:c065426c (nsched: 2913)
[Shell(6038:6038)]:   Sleeping       Pr:30 SyncTask(3:3)                  @ 0008:c065426c (nsched: 3663)
[Shell(6038:6038)]:   Mutex          Pr:10 FinalizerTask(4:4)             @ 0008:c065426c (nsched: 6798)
[Shell(6038:6038)]:   Sleeping       Pr:30 UHCIHotplug(5:5)               @ 0008:c065426c (nsched: 3639)
[Shell(6038:6038)]:   Selecting      Pr:50 AudioServer(12:12)             @ 001b:03c0e336 (nsched: 584)
[Shell(6038:6038)]:   Selecting      Pr:10 DHCPClient(13:13)              @ 001b:71c97336 (nsched: 519)
[Shell(6038:6038)]:   Futex          Pr:30 AudioServer(12:14)             @ 001b:03c0e336 (nsched: 4)
[Shell(6038:6038)]:   Selecting      Pr:30 Shell(22:22)                   @ 001b:5b16d336 (nsched: 739)
[Shell(6038:6038)]:   Waiting        Pr:30 run-tests(5801:5801)           @ 001b:5d93f336 (nsched: 154)
[Shell(6038:6038)]:   Selecting      Pr:30 Shell(6033:6033)               @ 001b:b044d336 (nsched: 149)
[Shell(6038:6038)]:   Running        Pr:30 Shell(6038:6038)               @ 001b:b044d336 (nsched: 5)
[Shell(6038:6038)]:   Runnable       Pr:30 head(6039:6039)                @ 001b:06492336 (nsched: 65)
[Shell(6038:6038)]:   Dead           Pr:30 echo(6040:6040)                @ 001b:3f122336 (nsched: 65)
[#0 Shell(6038:6038)]: PV(error) fault in Region(0xc0e9d0c0)[15] at V0xc6514bdf
[#0 Shell(6038:6038)]: Unrecoverable page fault, instruction fetch / read from address V0xc6514bdf
[Shell(6038:6038)]: CRASH: CPU #0 Page Fault in ring 0
[#0 Shell(6038:6038)]: Exception code: 0011 (isr: 0000)
[#0 Shell(6038:6038)]:     pc=0x0008:0xc6514bdf eflags=0x00000086
[#0 Shell(6038:6038)]:  stack=0x0010:0xc6514bb0
[#0 Shell(6038:6038)]:     ds=0x0010 es=0x0010 fs=0x0023 gs=0x0030
[#0 Shell(6038:6038)]:    eax=0x81ca2f10 ebx=0xc0df114c ecx=0xc0df3684 edx=0xc0df3684
[#0 Shell(6038:6038)]:    ebp=0xc6514bbc esp=0xc6514bb0 esi=0xc0df2018 edi=0xc63d5144
[#0 Shell(6038:6038)]:    cr0=0x80010013 cr2=0xc6514bdf cr3=0x34a8e000 cr4=0x000006e0
[Shell(6038:6038)]: KERNEL PANIC! :^(
[Shell(6038:6038)]: Crash in ring 0
[Shell(6038:6038)]: at ../../Kernel/Arch/x86/common/Interrupts.cpp:230 in void Kernel::handle_crash(const Kernel::RegisterSta
te &, const char *, int, bool)
[Shell(6038:6038)]: Kernel + 0x0034fc14  Kernel::__panic(char const*, unsigned int, char const*) +0x44
[Shell(6038:6038)]: Kernel + 0x004553d9  Kernel::handle_crash(Kernel::RegisterState const&, char const*, int, bool) +0x141
[Shell(6038:6038)]: Kernel + 0x00456339  page_fault_handler +0x8c4
[Shell(6038:6038)]: Kernel + 0x00455a70  page_fault_asm_entry +0x26
```

**Comment by ADKaster** [comment]
Ok this repros much faster just running `for i in {1..5} { /usr/Tests/Shell/sigpipe.sh }` a few times I got the deadlock on the third run.

Interesting to note that doing `for i in {1..20} { /usr/Tests/Shell/sigpipe.sh & }` locks up the kernel so badly that you can't get the gdb interrupt to pause the processor.

**Comment by ADKaster** [comment]
I'm really not sure what's going on here.. So ... *in the constructor* for `WaitBlocker` and/or WaitBlockCondition, it can't get its own spinlock because it's already 1?

This seriously smells like a "use before fully constructed" bug.

I was looking at Thread::block<BlockType> and it takes .. a lot of locks. 

https://github.com/SerenityOS/serenity/blob/0152f1924b9330e00c027c8b0cc0e7cab22a89f0/Kernel/Thread.h#L860-L869

I thought that maybe this is happeing:
-  we're taking the block lock spinlock for a thread, from waitid (Thread::block)
- we construct our WaitBlocker (ThreadBlockers.cpp:592)
- Somehow, we end up getting unscheduled.
- ????
- Deadlock

So I moved the scheduler_lock up above the construction of the blocker, and found something funky

```
[Shell(330:330)]: ASSERTION FAILED: !g_scheduler_lock.own_lock()
[Shell(330:330)]: ../../Kernel/Thread.cpp:201 in void Kernel::Thread::block(Kernel::Mutex &, ScopedSpinLock<SpinLock<u8>> &,
u32)
[Shell(330:330)]: KERNEL PANIC! :^(

#0  Kernel::halt_this () at ../.././Kernel/Arch/x86/ASM_wrapper.h:146
#1  Kernel::Processor::halt () at ../../Kernel/Arch/x86/common/Processor.cpp:1021
#2  0xc0552b60 in Kernel::__panic (file=..., line=..., function=...) at ../../Kernel/Panic.cpp:33
#3  0xc065875d in abort () at ../../Kernel/Arch/x86/common/CPU.cpp:28
#4  0xc06586cc in __assertion_failed (msg=..., file=..., line=..., func=...) at ../../Kernel/Arch/x86/common/CPU.cpp:18
#5  0xc061e0fb in Kernel::Thread::block (this=..., lock=..., lock_lock=..., lock_count=...) at ../../Kernel/Thread.cpp:201
#6  0xc04ebf68 in Kernel::Mutex::block (this=..., current_thread=..., mode=..., lock=..., requested_locks=...)
    at ../../Kernel/Locking/Mutex.cpp:206
#7  0xc04eaf71 in Kernel::Mutex::lock (this=..., mode=..., location=...) at ../../Kernel/Locking/Mutex.cpp:52
#8  0xc0387922 in Kernel::MutexLocker::MutexLocker (this=..., l=..., mode=..., location=...)
    at ../.././Kernel/Locking/Mutex.h:117
#9  Kernel::LockedResource<AK::IntrusiveListRelaxedConst<Kernel::Process, Kernel::Process*, &Kernel::Process::m_list_node> const, (Kernel::LockMode)1>::LockedResource (this=..., value=..., mutex=..., location=...)
    at ../.././Kernel/Locking/ContendedResource.h:21
#10 0xc03876e6 in Kernel::ProtectedValue<AK::IntrusiveListRelaxedConst<Kernel::Process, Kernel::Process*, &Kernel::Process::m_list_node> >::lock_shared (this=..., location=...) at ../.././Kernel/Locking/ProtectedValue.h:24
#11 0xc0556ea0 in Kernel::ProtectedValue<AK::IntrusiveListRelaxedConst<Kernel::Process, Kernel::Process*, &Kernel::Process::m_list_node> >::with_shared<Kernel::Process::from_pid(AK::DistinctNumeric<int, Kernel::__ProcessID_tag, false, true, true, false, false, false>)::$_4>(Kernel::Process::from_pid(AK::DistinctNumeric<int, Kernel::__ProcessID_tag, false, true, true, false, false, false>)::$_4, AK::SourceLocation const&) const (this=..., callback=..., location=...)
    at ../.././Kernel/Locking/ProtectedValue.h:35
#12 Kernel::Process::from_pid (pid=...) at ../../Kernel/Process.cpp:421
#13 0xc0635179 in Kernel::Thread::WaitBlocker::WaitBlocker (this=..., wait_options=..., id_type=..., id=..., result=...) at ../../Kernel/ThreadBlockers.cpp:601
#14 0xc060177e in Kernel::Thread::block<Kernel::Thread::WaitBlocker, int&, idtype_t&, int&, Kernel::KResultOr<siginfo>&> (this=..., timeout=..., args=..., args=..., args=..., args=...) at ../.././Kernel/Thread.h:868
#15 0xc0601041 in Kernel::Process::do_waitid (this=..., idtype=..., id=..., options=...) at ../../Kernel/Syscalls/waitid.cpp:15
#16 0xc06012b4 in Kernel::Process::sys$waitid (this=..., user_params=...) at ../../Kernel/Syscalls/waitid.cpp:41
#17 0xc058ab6d in Kernel::Syscall::handle (regs=..., function=..., arg1=..., arg2=..., arg3=..., arg4=...) at ../../Kernel/Syscall.cpp:159
#18 Kernel::syscall_handler (trap=...) at ../../Kernel/Syscall.cpp:216
#19 0xc058a461 in Kernel::syscall_asm_entry () at ../../Kernel/Syscall.cpp:26
```

While we're trying to add a blocker to the thread for waitid, we call WaitBlocker::WaitBlocker, which then calls Process::from_pid, which *takes a Mutex* and then tries to block the thread again! I think this is our culprit right here, trying to recursively lock a spinlock in block<BlockCondition>()

@tomuta  what do you think of this analysis? This Blocker stuff is kinda hard to follow.. I remember you mentioning some kinda WIP rework for it :D

**Comment by tomuta** [comment]
Yeah so this isn't going to work, the process list can't use a mutex, it has to be a spinlock. Basically, the `WaitBlocker`, and more specifically the base class `Thread::Blocker` have the `m_block_lock` held when being created, and there isn't really a way around that. `m_block_lock` also can't be a `Mutex` for obvious reasons.

I don't have any rework or anything planned in that area, but the removal of the `g_scheduler_lock` also allows us to remove the `Thread::m_block_lock` and uses `Thread::m_lock` instead, but since that's also a spinlock (and it has to be) it wouldn't solve this problem.

The problem really is that it's not legal to use or acquire a `Mutex` while any `SpinLock` is being held.

We really need to add a `VERIFY_INTERRUPTS_ENABLED()` to `Mutex::lock` and `Mutex::unlock` to catch this sort of thing not only for the case where preemption would actually be necessary. We should be able to do this, now that critical sections and spinlocks behave more correctly.

**Comment by tomuta** [comment]
Yeah, adding `VERIFY_INTERRUPTS_ENABLED()` to both `Mutex::lock` and `Mutex::unlock` makes the system blow up quite early on for the same reasons. We need to fix all of those.

**Comment by ADKaster** [comment]
That VERIFY_INTERRUPTS_ENABLED() suggestion seems to be quite the hairy yak. after changing the all processes list into a SpinlockProtected list, we do interesting things like take FS mutexes from the page fault handler.

```
[init_stage2(1:1)]: ASSERTION FAILED: cpu_flags() & 0x200
[init_stage2(1:1)]: ../../Kernel/Locking/Mutex.cpp:21 in void Kernel::Mutex::lock(Kernel::Mutex::Mode, const Kernel::LockLocation&)
[init_stage2(1:1)]: KERNEL PANIC! :^(
[init_stage2(1:1)]: Aborted
[init_stage2(1:1)]: at ../../Kernel/Arch/x86/common/CPU.cpp:28 in void abort()
[init_stage2(1:1)]: Kernel + 0x0046821d  Kernel::__panic(char const*, unsigned int, char const*) +0xb8
[init_stage2(1:1)]: Kernel + 0x0061cc84  illegal_instruction_asm_entry +0x0
[init_stage2(1:1)]: Kernel + 0x0061caa5  abort.localalias +0x0
[init_stage2(1:1)]: Kernel + 0x003cd03e  Kernel::Mutex::lock(Kernel::LockMode, Kernel::LockLocation const&) +0x70
[init_stage2(1:1)]: Kernel + 0x00281f69  Kernel::Ext2FSInode::read_bytes(long long, unsigned long, Kernel::UserOrKernelBuffer&, Kernel::FileDescription*) const [clone .localalias] +0x117
[init_stage2(1:1)]: Kernel + 0x003b6b08  Kernel::Memory::Region::handle_inode_fault(unsigned long) [clone .localalias] +0x966
[init_stage2(1:1)]: Kernel + 0x003b8ae2  Kernel::Memory::Region::handle_fault(Kernel::PageFault const&) +0x476
[init_stage2(1:1)]: Kernel + 0x0039ad9a  Kernel::Memory::MemoryManager::handle_page_fault(Kernel::PageFault const&) +0x27e
[init_stage2(1:1)]: Kernel + 0x0061f8b1  page_fault_handler +0x5f2
[init_stage2(1:1)]: Kernel + 0x0061cd64  page_fault_asm_entry +0x26
[init_stage2(1:1)]: Kernel + 0x00683cdc  ELF::Image::parse() [clone .localalias] +0x148
[init_stage2(1:1)]: Kernel + 0x0068464c  ELF::Image::Image(unsigned char const*, unsigned long, bool) +0x6a
[init_stage2(1:1)]: Kernel + 0x004d4ba3  Kernel::load_elf_object(AK::NonnullOwnPtr<Kernel::Memory::AddressSpace>, Kernel::FileDescription&, unsigned int, Kernel::ShouldAllocateTls, Kernel::ShouldAllowSyscalls) +0x66c
[init_stage2(1:1)]: Kernel + 0x004d8ba1  Kernel::Process::load(AK::NonnullRefPtr<Kernel::FileDescription>, AK::RefPtr<Kernel::FileDescription, AK::RefPtrTraits<Kernel::FileDescription> >, elfhdr const&) [clone .localalias] +0x1011
[init_stage2(1:1)]: Kernel + 0x004dbff0  Kernel::Process::do_exec(AK::NonnullRefPtr<Kernel::FileDescription>, AK::Vector<AK::String, 0ul>, AK::Vector<AK::String, 0ul>, AK::RefPtr<Kernel::FileDescription, AK::RefPtrTraits<Kernel::FileDescription> >, Kernel::Thread*&, u
nsigned int&, elfhdr const&) [clone .localalias] +0xefa
[init_stage2(1:1)]: Kernel + 0x004e3d00  Kernel::Process::exec(AK::String, AK::Vector<AK::String, 0ul>, AK::Vector<AK::String, 0ul>, int) [clone .part.0] +0x1606
[init_stage2(1:1)]: Kernel + 0x004e4751  Kernel::Process::exec(AK::String, AK::Vector<AK::String, 0ul>, AK::Vector<AK::String, 0ul>, int) [clone .localalias] +0x71
[init_stage2(1:1)]: Kernel + 0x00485d91  Kernel::Process::create_user_process(AK::RefPtr<Kernel::Thread, AK::RefPtrTraits<Kernel::Thread> >&, AK::String const&, unsigned int, unsigned int, AK::DistinctNumeric<int, Kernel::__ProcessID_tag, false, true, true, false, fal
se, false>, int&, AK::Vector<AK::String, 0ul>&&, AK::Vector<AK::String, 0ul>&&, Kernel::TTY*) +0x1a99
[init_stage2(1:1)]: Kernel + 0x0060fd0f  Kernel::init_stage2(void*) +0x17f6
[init_stage2(1:1)]: Kernel + 0x006223ec  exit_kernel_thread +0x0
```

**Comment by ADKaster** [comment]
I would assume that we should instead do something like block the offending thread until the bytes can be fetched from disk, and then switch context to someone else? Assuming that's a valid thing to do for an inode fault...

**Comment by ADKaster** [comment]
Also SyncTask apparently?
```
[SyncTask(3:3)]: ASSERTION FAILED: cpu_flags() & 0x200
[SyncTask(3:3)]: ../../Kernel/Locking/Mutex.cpp:205 in void Kernel::Mutex::block(Kernel::Thread&, Kernel::Mutex::Mode, Kernel::SpinlockLocker<Kernel::Spinlock<unsigned char> >&, u32)
[SyncTask(3:3)]: KERNEL PANIC! :^(
[SyncTask(3:3)]: Aborted
[SyncTask(3:3)]: at ../../Kernel/Arch/x86/common/CPU.cpp:28 in void abort()
[SyncTask(3:3)]: Kernel + 0x004682f7  Kernel::__panic(char const*, unsigned int, char const*) +0xb8
[SyncTask(3:3)]: Kernel + 0x0061cd98  illegal_instruction_asm_entry +0x0
[SyncTask(3:3)]: Kernel + 0x0061cbb9  abort.localalias +0x0
[SyncTask(3:3)]: Kernel + 0x003cb22f  Kernel::Mutex::block(Kernel::Thread&, Kernel::LockMode, Kernel::SpinlockLocker<Kernel::Spinlock<unsigned char> >&, unsigned int) [clone .localalias] +0x85
[SyncTask(3:3)]: Kernel + 0x003cd847  Kernel::Mutex::lock(Kernel::LockMode, Kernel::LockLocation const&) +0x80d
[SyncTask(3:3)]: Kernel + 0x00246200  Kernel::BlockBasedFileSystem::flush_writes_impl() [clone .localalias] +0x1a2
[SyncTask(3:3)]: Kernel + 0x002962f8  Kernel::Ext2FS::flush_writes() [clone .localalias] +0x129e
[SyncTask(3:3)]: Kernel + 0x002b6887  Kernel::FileSystem::sync() +0x98b
[SyncTask(3:3)]: Kernel + 0x00336333  Kernel::VirtualFileSystem::sync() +0x15
[SyncTask(3:3)]: Kernel + 0x005c483a  void Kernel::Process::kernel_process_trampoline<Kernel::SyncTask::spawn()::{lambda()#1}>(void*) +0x99
```

**Comment by ADKaster** [comment]
```
        InterruptDisabler disabler;
        if (auto shared_vmobject = inode.shared_vmobject()) {
            if ((prot & PROT_EXEC) && shared_vmobject->writable_mappings())
                return false;
            if ((prot & PROT_WRITE) && shared_vmobject->executable_mappings())
                return false;
        }
```
:thonk:

**Comment by tomuta** [comment]
Yeah those `InterruptDisabler` won't be helpful on SMP either... But I think acquiring (and potentially blocking due to) a `Mutex` in a page fault handler probably should be legal.

**Comment by ADKaster** [comment]
Haven't seen the deadlock in a while, only the write_cr3 crash.
